planner/core/rule_join_reorder_dp.go

```diff
@@ -0,0 +1,181 @@
+			visited[next] = true
+		}
+	}
+	return newPos2OldPos
+}
+
+func (s *joinReorderDPSolver) dpGraph(newPos2OldPos, oldPos2NewPos []int, joinGroup []LogicalPlan, totalEdges []joinGroupEdge) (LogicalPlan, error) {
+	nodeCnt := len(newPos2OldPos)
+	bestPlan := make([]LogicalPlan, 1<<uint(nodeCnt))
+	bestCost := make([]int64, 1<<uint(nodeCnt))
+	// bestPlan[s] is nil can be treated as bestCost[s] = +inf.
+	for i := 0; i < nodeCnt; i++ {
+		bestPlan[1<<uint(i)] = joinGroup[newPos2OldPos[i]]
+	}
+	// Enumerate the state from small to big, make sure that S1 must be enumerated before S2 if S1 belongs to S2.
+	for state := uint(1); state < (1 << uint(nodeCnt)); state++ {
+		if bits.OnesCount(state) == 1 {
+			continue
+		}
+		// This loop can iterate all its subset.
+		for sub := (state - 1) & state; sub > 0; sub = (sub - 1) & state {
+			remain := state ^ sub
+			if sub > remain {
+				continue
+			}
+			// If this subset is not connected skip it.
+			if bestPlan[sub] == nil || bestPlan[remain] == nil {
+				continue
+			}
+			// Get the edge connecting the two parts.
+			usedEdges := s.nodesAreConnected(sub, remain, oldPos2NewPos, totalEdges)
+			if len(usedEdges) == 0 {
+				continue
+			}
+			join, err := s.newJoinWithEdge(bestPlan[sub], bestPlan[remain], usedEdges)
+			if err != nil {
+				return nil, err
+			}
+			if bestPlan[state] == nil || bestCost[state] > join.statsInfo().Count()+bestCost[remain]+bestCost[sub] {
+				bestPlan[state] = join
+				bestCost[state] = join.statsInfo().Count() + bestCost[remain] + bestCost[sub]
+			}
+		}
+	}
+	return bestPlan[(1<<uint(nodeCnt))-1], nil
+}
+
+func (s *joinReorderDPSolver) nodesAreConnected(leftMask, rightMask uint, oldPos2NewPos []int, totalEdges []joinGroupEdge) []joinGroupEdge {
+	var usedEdges []joinGroupEdge
+	for _, edge := range totalEdges {
+		lIdx := uint(oldPos2NewPos[edge.nodeIDs[0]])
+		rIdx := uint(oldPos2NewPos[edge.nodeIDs[1]])
+		if (leftMask&(1<<lIdx)) > 0 && (rightMask&(1<<rIdx)) > 0 {
+			usedEdges = append(usedEdges, edge)
+		} else if (leftMask&(1<<rIdx)) > 0 && (rightMask&(1<<lIdx)) > 0 {
+			usedEdges = append(usedEdges, edge)
+		}
+	}
+	return usedEdges
+}
+
+func (s *joinReorderDPSolver) newJoinWithEdge(leftPlan, rightPlan LogicalPlan, edges []joinGroupEdge) (LogicalPlan, error) {
+	var eqConds []*expression.ScalarFunction
+	for _, edge := range edges {
+		lCol := edge.edge.GetArgs()[0].(*expression.Column)
+		rCol := edge.edge.GetArgs()[1].(*expression.Column)
+		if leftPlan.Schema().Contains(lCol) {
+			eqConds = append(eqConds, edge.edge)
+		} else {
+			newSf := expression.NewFunctionInternal(s.ctx, ast.EQ, edge.edge.GetType(), rCol, lCol).(*expression.ScalarFunction)
+			eqConds = append(eqConds, newSf)
+		}
+	}
+	join := s.newJoin(leftPlan, rightPlan, eqConds)
+	_, err := join.deriveStats()
+	return join, err
+}
+
+// Make cartesian join as bushy tree.
+func (s *joinReorderDPSolver) makeBushyJoin(cartesianJoinGroup []LogicalPlan) LogicalPlan {
```

Comment:
Seems like it is the same as `joinReorderSolver.makeBushyJoin`. Also, do we need to optimize the cartesian join order?

Reply:
It's the same, yes. We can extract it. This can be done when addressing other comment.
We don't need to optimize that case. But when we extract the join group, we can not skip the cartesian one easily.
 e.g. `select ... from t1,t2,t3,t4 where t1.a=t3.a and t3.b=t4.b`. `t1, t2, t3, t4` is a join group, t2 is a cartesian one, we don't need to consider it. But the initial join tree may be `{{{t1, t2}, t3}, t4}` so we would add it to our join group too.